### ISSUE TITLE
no suitable port and timeout error when starting multile brokers

### ISSUE THREAD

**Issue by nguyenxuantuong** [mesos/kafka#127]
Hi,

I am running the mesos-kafka in one server within 3 mesos+zookeeper servers. But i am not sure about which ports I need to open between those servers. If I allow every ports to be opens for the there server, the system works as expected. But if I close the ports then I got timeout errors. I have few questions:
1. It is good practice to set specific port for each brokers using ./kafka-mesos.sh broker update 2 --port=31092. What is port range I can use. I tried port = 9091 but it's not working? It has to be > 30000?
2. I set three brokers with 3 different port: 31090->31092. But only 2 working, 1 got timeout. I need to choose 3 different ports for 3 brokers right?
3. If I want to reset the port to default "auto", how I can do it? I've tried ./kafka-mesos.sh broker update 2 --port="" but it's not working
4. What would be the broker endpoint I should use if I have external code to produce and consume message. Previously, I am using: server1:9091,server2:9091,server3:9091

Thanks you for the support

**Comment by joestein** [comment]
1) I often suggest this approach of setting the static port for different brokers with different concerns. This allows your code to have a known port for different clusters that have different responsibilities (i.e. logging app data vs transactions for upstream live traffic vs analytics brokers in an aggregate dc). Then you can use MesosDNS and not have to worry about any service discovery to get to your kafka cluster it is a known end point for the different clusters different needs. The range is determined by what is being advertised on the agent. The default ports the slaves offer are 31000-32000 you can change that in the agent resource file.

2) Well, if you want to run more than one kafka broker on a machine you have to give them different ports unless you use https://github.com/mesosphere/net-modules for floating IP address for agents. I don't know what timeout you got but if you have 3 agents you can use the same IP.

3) You have to update the data in the either the local json file of the scheduler or the zookeeper znode and remove the "port" key/value from it. There is no way to-do that in the api but would make a good pull request if you wanted to give it a shot? maybe port=0 and in the condition where it checks if new BindAddress or null add condition for 0 and update readme.

4) you can query api/broker/status from scheduler and parse all the end points or through http://mesosphere.github.io/mesos-dns/

**Comment by nguyenxuantuong** [comment]
Hi,

Thank you so much for your detailed answer.

I was able to config three broker running by calling "broker add" and "broker start" on each of 3 different machines instead of running on a single one. But I got another problem.

I wanted to delete all of the newly created topics. So what I did is remove all of the broker and delete everything in the logs.dir (which doesn't seem to be a good way to do). But after that, i was no longer able to start a new broker again. I always got Error: broker start timeout no matter how hard I tried. I tried to stop all mesos-master, mesos-slave, and delete all data in zookeeper, but after restarting mesos, I still see something like:

Ignoring status update TASK_LOST (UUID: c223a35d-7807-43af-ac14-570b99729fe3) for task broker-1-181b2793-fb8b-430d-9f26-71a3346a4ee7 of framework 20151009-155844-50393098-5050-19170-0000 from slave 20151009-155844-50393098-5050-19170-S112 at slave(1)@10.240.0.5:5051 (zoo3) because the framework is unknown

And ./kafka-mesos.sh broker start 0 keep getting timeout. It seems that the broker-1 task which was created previously has not been removed completely.

Do you know how to completely remove all of the orphan task in mesos (keep it as fresh installed) and is there any idea why I was no longer able to readd and start a broker?

It's very frustrated to keep getting timeout error. Hope that you can support me on that.

Thanks.

PS: I have three servers. I have installed zookeeper on all there and running mesos-master and mesos-slave in all there server. The kafka-mesos-scheduler is run on one servers; "broker add" is run in each server.

**Comment by dmitrypekar** [comment]
Hi @nguyenxuantuong 

Let me clarify a little bit.

First, to reset the port you should use `./kafka-mesos.sh update $id --port ""` instead of `./kafka-mesos.sh update $id --port=""`.

Second, about TASK_LOST.
Those status means, that for some reason Mesos is failing to start the Executor running Kafka Broker. The could be dozens of different reason why this could happen. The precise reason of that could be found in `stdout` and `stderr` files in the Sandbox directory. The Sandbox directory could be browsed using Mesos web UI. 

Could you please attach `stdout` and `stderr` files for us to be able to find the root cause of the problem?

**Comment by nguyenxuantuong** [comment]
Hi @dmitrypekar,

Thank you for your answer. By taking a look at the logs from WebUI, I was able to identify the problem. It was because I ran the scheduler using an upstart script (by right it should be ran on Marathon) but unfortunately the upstart script is unable to identify the MESOS_NATIVE_JAVA_LIBRARY environment variable. After setting environment parameters in the upstart script, everything start to work correctly.

Thanks.
